homeassistant/components/sensor/filter.py

```diff
@@ -277,6 +288,47 @@ def _filter_state(self, new_state):
         return filtered
 
 
+@FILTERS.register(FILTER_NAME_TIME_SMA)
+class TimeSMAFilter(Filter):
+    """Simple Moving Average (SMA) Filter.
+
+    The window_size is determined by time, and SMA is time weighted.
+    """
+
+    def __init__(self, window_size, precision, entity):
+        """Initialize Filter."""
+        super().__init__(FILTER_NAME_TIME_SMA, 0, precision, entity)
+        self._time_window = int(window_size.total_seconds())
+        self.last_leak = None
+        self.queue = deque()
+
+    def _leak(self):
+        """Remove timeouted elements."""
+        now = int(datetime.now().timestamp())
+
+        while self.queue:
+            timestamp, _ = self.queue[0]
+            if timestamp + self._time_window <= now:
+                self.last_leak = self.queue.popleft()
+            else:
+                break
+        return now
+
+    def _filter_state(self, new_state):
+        now = self._leak()
+        self.queue.append((now, float(new_state)))
+        moving_sum = 0
+        start = now - self._time_window
+        _, prev_val = self.last_leak or (0, float(new_state))
+
+        for timestamp, val in self.queue:
+            moving_sum += (timestamp-start)*prev_val
+            start, prev_val = timestamp, val
+        moving_sum += (now-start)*prev_val
+
+        return moving_sum/self._time_window
```

Comment:
What if we don't receive any value within `self._time_window`? I think we should be reporting an `unknown` state then. For example if my time window is 1 minute but I haven't received any state updates for 2 minutes, the output of this sensor should automatically be set to `unknown` (and not just repeat the same old state indefinitely).

I know this might be a bit of work to do but it's what I'd expect as a user. See usages of `async_track_point_in_utc_time` for ways how to do this.

Reply:
I disagree, If I don't get any values it just means that it hasn't changed.

Reply:
The problem here is that the time has changed and time is also an input into this filter. Imagine that our current value is based off 2 states that happened during the window. If we wait long enough that 1 state drops out of our window, wouldn't you expect to base the value off just the other state?

Reply:
@balloob that is exactly what it will happen.

Even if no points are left in the time window, the filter will report based on the last value that exited the windows (previous_val)

Reply:
@dgomes I think you're misunderstanding Paulus' comment (or I'm not understanding it...)

If there's **only one** point within the time window, as a user I wouldn't expect it to use the last value outside the time window for any computation. It should just discard anything that's outside the time window. See this gif for an example (the blue points are the data points, the blue area is our time window and red line is the output):

![out](https://user-images.githubusercontent.com/6833237/37484641-685f36ac-2889-11e8-9357-c4eb02409cff.gif)

Couple of things going on here:

* First we only "see" the first value, so of course we should only report that value.
* Then we start to see the next value. We should then take the weighted average (or whatever it's called in your paper) of the two.
* If the first value exits the window, we should not use it for any computation anymore. In your current code that is however happening, not what I would expect.
* *Optional:* In my opinion if there's no data points inside our window left, as I said we should report an unknown state.

This might not be exactly the time_sma from the paper but again we should not depend on some exact definition but rather on what users would expect us to do.

Reply:
how do you make these fancy graphs?

Reply:
@OttoWinter that is not the filter I'm implementing here. 

Sure, that filter can be implemented, but it's not the one in this PR.

Please share how you are creating those graphs 😄

Reply:
@dgomes I see that you report only on the last value but you only update the filter value when a new state comes in right?

Reply:
@dgomes Ok, sorry I will shut up about other filters 😅

Yeah those graphics were done in [GeoGebra](https://www.geogebra.org/), but I would describe it as the single most user-unfriendly maths tools ever written IMHO (but it's sadly what's taught here in Austria 🇦🇹). Seriously, it's incredibly hard to use if you're starting out. (incidentally, because of its poor usability the rest of my class will probably also have to re-take a recent maths exam 😥)

btw, if for some reason you're looking for a nice tool (in python, yay) for creating nice presentations, have a look at [manim](https://github.com/3b1b/manim). For example, it has allowed me to create these fancy visualizations about graph theory: https://youtu.be/t2Gb1YA-HcU 😺

Reply:
@balloob right, this filter always lags 1 state behind. The concept is that the raw sensor has maintained the previous status until the new state has arrived. 

**E.g.**: Think about a water pump that publishes its rate (1 to 5). If the pump is at speed 5 and comes down to speed 4 (new state) the SMA is calculated with the period of time in which the pump was at 5 (we weight in the time the pump was at 5), 4 is of no use in the SMA since the that period of time is 0 (it just got to 4). 

@OttoWinter please contribute more filters :) It's just that I'm prioritising filters I can use in my setup. There are plenty of important filters missing which I hope some else can contribute 😄

Reply:
Okay I'm going to shut up talking about filters as I have clearly not a clue what I am talking about 👍